monkey/monkey_island/cc/services/agent_signals_service.py

```diff
@@ -28,18 +28,19 @@ def get_signals(self, agent_id: AgentID) -> AgentSignals:
         return AgentSignals(terminate=terminate_timestamp)
 
     def _get_terminate_signal_timestamp(self, agent_id: AgentID) -> Optional[datetime]:
```

Comment:
The purpose of this method is to determine the terminate signal timestamp. The changes added have nothing to do with the terminate signal timestamp, because in the case of heartbeat, there was no terminate signal sent. Extract the changes into a separate method, like `get_terminated_by_heartbeat_timestamp` or something. Then override the terminate signal timestamp with that.

Reply:
I think you're assuming that there's one single terminate signal for the system. This is true in practice, but the user of `_get_terminate_signal_timestamp()` does not make this assumption. Reason: This could change in the future.

`_get_terminate_signal_timestamp()` gets terminate signals **per agent**. In the future, we could have the heartbeat handler change the terminate signal time for a specific agent, but we're not set up for that yet. This logic should stay where it is.

Reply:
I'm not assuming that. This method handles the logic of the latest, global shutdown signal. We could rename it to `_get_global_terminate_signal_timestamp()`. Heartbeat shutdown happens for specific agents, it has nothing to do with the global shutdown, so the logic should be separate. There might be many signals in the future, like global shutdown, per-agent shutdown and heartbeat shutdown. Separate methods might fetch those times and then return the latest to the user, but merging everything into a single method is not a good solution.

Reply:
Fixed by https://github.com/guardicore/monkey/pull/2654/commits/af7399ee4e7187a22341c0479fb69da56ab56ca0